runtime/memory.c

```diff
@@ -498,7 +507,7 @@ static inline value caml_alloc_shr_aux (mlsize_t wosize, tag_t tag,
       else if (caml_in_minor_collection)
```

Comment:
This is the line that makes `Out_of_memory` unreliable. You should remove the test and always call `caml_fatal_error` in this case.